src/safepoints.rs

```diff
@@ -0,0 +1,138 @@
+extern crate ykstackmaps;
+
+use std::collections::HashMap;
+use std::path::Path;
+use ykstackmaps::{LocKind, LocOffset, SMRec, StackMapParser};
+
+use core::mem;
+
+#[derive(Clone, Copy, Hash, PartialEq, Eq, Debug)]
+pub struct ReturnAddress(pub u64);
+
+/// The offset from the stack pointer.
+#[derive(Debug)]
+struct SPO(u32);
+
+/// A `PtrSlot` identifies a stack root at a given safepoint using its offset
+/// from the Stack Pointer.
+///
+/// A base pointer (not to be confused with X86 terminology, where base pointer
+/// refers to the frame pointer register) is a pointer an object. In opposition
+/// to this, a derived pointer points to the interior of an object.
+///
+/// The Derived variant of a `PtrSlot` also contains a Stack Pointer offset to
+/// the base of the object.
+#[derive(Debug)]
+enum PtrSlot {
+    Base(SPO),
+    Derived(SPO, SPO),
+}
+
+/// Contains root locations for a Safepoint
+#[derive(Debug)]
+pub struct SafepointRoots {
+    /// A list of registers which contain roots across a safepoint
+    // DWARF Register number mapping can be found here:
+    // Pg.63 https://software.intel.com/sites/default/files/article/402129/mpx-linux64-abi.pdf
+    registers: Vec<u16>,
+
+    /// A list of `PtrSlot`s which correspond to roots accessible from a stack
+    /// pointer offset across a safepoint.
+    stack_offsets: Vec<PtrSlot>,
+}
+
+/// Pointer slots in stackmap locations are treated as offsets from a
+/// particular register. In the case of a Direct or Indirect location kind,
+/// these are the frame pointer and stack pointer registers respectively. To
+/// avoid calculating this during a GC pause, we convert all offsets to be
+/// from an SP upfront.
+fn as_sp_offset(offset: &LocOffset) -> SPO {
+    match offset {
+        LocOffset::I32(ref o) => SPO(*o as u32),
+        _ => panic!("Offset must be signed"),
+    }
+}
+
+fn gen_safepoint_roots(stackmap: SMRec) -> SafepointRoots {
+    // The first 2 locations are uninteresting, however, they should be constants.
+    let mut idx = 2;
```

Comment:
This feels like it should probably go down to about line 70 or so?

Also, does this location have a name that we can make a constant? In general, it's better to have named constants for this sort of thing.

Reply:
Even if it doesn't, I think we'd benefit from a constant with a home-grown name, e.g. `NUM_SKIP_STACKMAPS`.

Reply:
I agree, I'll fix this